### PR TITLE
feat(webview): add batchNearby utility for smarter tool ask batching

### LINKED ISSUE #1004: UI tool ask batching fails when models insert API rows between tool calls
# UI tool ask batching fails when models insert API rows between tool calls

## Problem

When using models like qwen, the UI shows multiple separate prompts for the same type of tool call:

```
Zoo wants to read this file (a.ts)
Zoo wants to read this file (b.ts)
Zoo wants to edit this file (c.ts)
```

Instead of a single batched prompt:

```
Zoo wants to read these files:
- a.ts
- b.ts
```

## Root Cause

The existing [`batchConsecutive()`](webview-ui/src/utils/batchConsecutive.ts) utility only merges tool asks that are **truly adjacent** in the message array. However, models like qwen insert low-information messages between tool calls during streaming:

- `say: "api_req_started"` — API request metadata row
- `say: "api_req_finished"` — API request completion row  
- `say: "text"` with empty content — partial streaming rows
- `say: "reasoning"` — hidden reasoning text

These messages break the consecutive chain, so `batchConsecutive` stops merging.

## Current Code (ChatView.tsx ~line 1272)

```ts
const readFileBatched = batchConsecutive(filtered, isReadFileAsk, synthesizeReadFileBatch)
const listFilesBatched = batchConsecutive(readFileBatched, isListFilesAsk, synthesizeListFilesBatch)
const result = batchConsecutive(listFilesBatched, isEditFileAsk, synthesizeEditFileBatch)
```

## Proposed Solution: `batchNearby()` utility

A new utility that merges same-type tool asks even when separated by ignorable messages. It uses three predicates:

1. **`isTarget`** — matches the target tool type (readFile, listFiles, editFile)
2. **`isIgnorableBetweenTargets`** — skips over low-info messages without breaking the batch
3. **`isBoundary`** — stops merging when hitting a semantic boundary

### Ignorable Messages (skipped during batching)
- `api_req_started` / `api_req_finished`
- Empty text rows (`say: "text"` with no content)
- Reasoning rows (`say: "reasoning"`)

### Semantic Boundaries (stop merging)
- User feedback (`user_feedback`, `user_feedback_diff`)
- Visible assistant text (`text` with content)
- Completion result (`completion_result`)
- Checkpoint saved (`checkpoint_saved`)
- Errors (`error`)
- Context condensation (`condense_context`)
- Codebase search results (`codebase_search_result`)

### New Usage in ChatView.tsx

```ts
const readFileBatched = batchNearby(filtered, {
  isTarget: isReadFileAsk,
  isIgnorableBetweenTargets, // api_req_started/finished, empty text, reasoning
  isBoundary,                // user_feedback, visible text, completion_result, error, etc.
  synthesize: synthesizeReadFileBatch,
})
```

## Files Changed

- **New**: `webview-ui/src/utils/batchNearby.ts` — the batching utility function
- **New**: `webview-ui/src/utils/__tests__/batchNearby.spec.ts` — 23 comprehensive tests
- **Modified**: `webview-ui/src/components/chat/ChatView.tsx` — replace `batchConsecutive` with `batchNearby`

## Testing

All 23 unit tests pass, covering:
- Empty input / no matches / single match passthrough
- Consecutive vs non-consecutive matching
- Ignorable messages between targets (api_req_started/finished, empty text, reasoning)
- Semantic boundaries stopping the merge (user_feedback, visible text, completion_result, checkpoint_saved, error)
- Multiple batches separated by boundaries
- Realistic qwen scenarios with API rows between tool calls

## Long-term Improvement (separate issue)

Add `toolCallGroupId` metadata to backend messages so UI can use explicit grouping instead of heuristic-based merging. This would make batching work consistently across all model providers regardless of streaming format differences.

### PR DESCRIPTION
**Pull request by easonLiangWorldedtech**
# UI tool ask batching fails when models insert API rows between tool calls

## Problem

When using models like qwen, the UI shows multiple separate prompts for the same type of tool call:

```
Zoo wants to read this file (a.ts)
Zoo wants to read this file (b.ts)
Zoo wants to edit this file (c.ts)
```

Instead of a single batched prompt:

```
Zoo wants to read these files:
- a.ts
- b.ts
```

## Root Cause

The existing [`batchConsecutive()`](webview-ui/src/utils/batchConsecutive.ts) utility only merges tool asks that are **truly adjacent** in the message array. However, models like qwen insert low-information messages between tool calls during streaming:

- `say: "api_req_started"` — API request metadata row
- `say: "api_req_finished"` — API request completion row  
- `say: "text"` with empty content — partial streaming rows
- `say: "reasoning"` — hidden reasoning text

These messages break the consecutive chain, so `batchConsecutive` stops merging.

## Current Code (ChatView.tsx ~line 1272)

```ts
const readFileBatched = batchConsecutive(filtered, isReadFileAsk, synthesizeReadFileBatch)
const listFilesBatched = batchConsecutive(readFileBatched, isListFilesAsk, synthesizeListFilesBatch)
const result = batchConsecutive(listFilesBatched, isEditFileAsk, synthesizeEditFileBatch)
```

## Proposed Solution: `batchNearby()` utility

A new utility that merges same-type tool asks even when separated by ignorable messages. It uses three predicates:

1. **`isTarget`** — matches the target tool type (readFile, listFiles, editFile)
2. **`isIgnorableBetweenTargets`** — skips over low-info messages without breaking the batch
3. **`isBoundary`** — stops merging when hitting a semantic boundary

### Ignorable Messages (skipped during batching)
- `api_req_started` / `api_req_finished`
- Empty text rows (`say: "text"` with no content)
- Reasoning rows (`say: "reasoning"`)

### Semantic Boundaries (stop merging)
- User feedback (`user_feedback`, `user_feedback_diff`)
- Visible assistant text (`text` with content)
- Completion result (`completion_result`)
- Checkpoint saved (`checkpoint_saved`)
- Errors (`error`)
- Context condensation (`condense_context`)
- Codebase search results (`codebase_search_result`)

### New Usage in ChatView.tsx

```ts
const readFileBatched = batchNearby(filtered, {
  isTarget: isReadFileAsk,
  isIgnorableBetweenTargets, // api_req_started/finished, empty text, reasoning
  isBoundary,                // user_feedback, visible text, completion_result, error, etc.
  synthesize: synthesizeReadFileBatch,
})
```

## Files Changed

- **New**: `webview-ui/src/utils/batchNearby.ts` — the batching utility function
- **New**: `webview-ui/src/utils/__tests__/batchNearby.spec.ts` — 23 comprehensive tests
- **Modified**: `webview-ui/src/components/chat/ChatView.tsx` — replace `batchConsecutive` with `batchNearby`

## Testing

All 23 unit tests pass, covering:
- Empty input / no matches / single match passthrough
- Consecutive vs non-consecutive matching
- Ignorable messages between targets (api_req_started/finished, empty text, reasoning)
- Semantic boundaries stopping the merge (user_feedback, visible text, completion_result, checkpoint_saved, error)
- Multiple batches separated by boundaries
- Realistic qwen scenarios with API rows between tool calls

## Long-term Improvement (separate issue)

Add `toolCallGroupId` metadata to backend messages so UI can use explicit grouping instead of heuristic-based merging. This would make batching work consistently across all model providers regardless of streaming format differences.

Fixes #1004

## Screen Cap
image

## Summary by CodeRabbit

- **Improvements**
  - Updated chat streaming behavior to batch related tool requests more reliably when low-information rows appear between them.
  - Added clearer “boundary” handling to prevent mixing across key semantic moments (e.g., user feedback, visible assistant text, completion outputs, checkpoints, and errors).
- **Tests**
  - Added a comprehensive new test suite covering nearby batching rules, including ignorable rows, boundary stopping, multi-batch segmentation, and immutability.

### CODE REVIEW

**Comment by easonLiangWorldedtech** [comment]
@coderabbitai review

File: webview-ui/src/utils/__tests__/batchNearby.spec.ts [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+			isIgnorableBetweenTargets,
+			isBoundary,
+			synthesize: synthesizeBatch,
+		})
+		expect(result).toHaveLength(2) // bridge succeeded → pending consumed; [BATCH:match-1,match-2, "done"]
+		expect(result[0].text).toBe("BATCH:match-1,match-2")
+		expect(result[1].text).toBe("done")
+	})
+
+	test("non-ignorable non-target message stops batching", () => {
```
Comment (edelauna) [inline webview-ui/src/utils/__tests__/batchNearby.spec.ts]: This test has no ignorable message before `command_output`, so `pendingIgnorable` is empty when the `else break` fires — the restore path at `batchNearby.ts:68–71` is never actually exercised here. What happens with input like `[match-1, api_req_started, command_output, match-2]`? The `api_req_started` should be restored to the output but no test currently covers that shape.

File: webview-ui/src/components/chat/ChatView.tsx [hunk 2]
```diff
@@ -1268,10 +1268,64 @@ const ChatViewComponent: React.ForwardRefRenderFunction<ChatViewRef, ChatViewPro
...
+		// Messages that can be safely skipped over when batching tool asks.
+		// These are low-information or invisible messages that don't affect semantics:
+		// - api_req_started / api_req_finished (API request metadata rows)
+		// - empty text rows (partial streaming with no visible content)
+		// - reasoning rows (hidden from user by default)
+		const isIgnorableBetweenTargets = (msg: ClineMessage): boolean => {
+			if (msg.type !== "say") return false
+			return (
+				msg.say === "api_req_started" ||
+				msg.say === "api_req_finished" ||
```
Comment (edelauna) [inline webview-ui/src/components/chat/ChatView.tsx]: Is `api_req_finished` reachable here? `combineApiRequests` (applied when building `modifiedMessages`) absorbs every `api_req_finished` into its paired `api_req_started` or drops it as an orphan, and `visibleMessages` adds a second filter at lines 1047–1050. If that's right, this arm can never fire and could be removed.

File: webview-ui/src/utils/__tests__/batchNearby.spec.ts [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+	if (m.type !== "say") return false
+	return (
+		m.say === "user_feedback" ||
+		m.say === "user_feedback_diff" ||
+		(m.say === "text" && !!m.text?.trim()) ||
+		m.say === "completion_result" ||
+		m.say === "checkpoint_saved" ||
+		m.say === "error" ||
+		m.say === "condense_context"
+	)
```
Comment (edelauna) [inline webview-ui/src/utils/__tests__/batchNearby.spec.ts]: The production `isBoundary` in `ChatView.tsx` includes `codebase_search_result` but this fixture doesn't. Worth keeping them in sync — and a test with `[readFileAsk, codebase_search_result_msg, readFileAsk]` would confirm it stops the batch.

File: webview-ui/src/utils/batchNearby.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
+/**
+ * Batch tool asks that are near each other, allowing ignorable messages in between.
+ *
+ * Unlike `batchConsecutive` which only merges truly adjacent items, this function
```
Comment (edelauna) [inline webview-ui/src/utils/batchNearby.ts]: Since `ChatView.tsx` was the only caller of `batchConsecutive`, that file and its spec are now dead code. Can you delete them in this PR to avoid confusing future readers. Dead code will start being enforced via knip soon.

**Review by edelauna — CHANGES_REQUESTED** [review]

File: webview-ui/src/utils/__tests__/batchNearby.spec.ts (hunk 1 above)
Comment (easonLiangWorldedtech) [reply webview-ui/src/utils/__tests__/batchNearby.spec.ts]: This is already covered now. The regression test starts on [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:250) and uses exactly the shape you described:

- `match-1` at [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:252)
- pending ignorable `api_req_started` at [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:253)
- non-ignorable `command_output` at [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:254)
- `match-2` at [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:255)

The assertions verify the restore path by checking that `api_req_started` is preserved in the output before `command_output`: [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:263) through [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:267). This exercises the restore branch in [`batchNearby.ts`](webview-ui/src/utils/batchNearby.ts:68), where pending ignorable messages are pushed back when the bridge fails.

File: webview-ui/src/components/chat/ChatView.tsx (hunk 2 above)
Comment (easonLiangWorldedtech) [reply webview-ui/src/components/chat/ChatView.tsx]: Already addressed on this branch — the `isIgnorableBetweenTargets` arm in `ChatView` no longer treats `api_req_finished` as ignorable. It now only allows `api_req_started`, empty `text`, and `reasoning` rows: [`ChatView.tsx`](webview-ui/src/components/chat/ChatView.tsx:1276).

That means an `api_req_finished` row cannot be used as a bridge for batching tool asks. In practice it should still be unreachable there because `modifiedMessages` is built through `combineApiRequests(...)`, and `visibleMessages` also filters `api_req_finished` out before rendering: [`ChatView.tsx`](webview-ui/src/components/chat/ChatView.tsx:131), [`ChatView.tsx`](webview-ui/src/components/chat/ChatView.tsx:1047).

File: webview-ui/src/utils/__tests__/batchNearby.spec.ts (hunk 3 above)
Comment (easonLiangWorldedtech) [reply webview-ui/src/utils/__tests__/batchNearby.spec.ts]: Already covered. The test fixture `isBoundary` now includes `codebase_search_result` on [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:43), matching the production boundary predicate in [`ChatView.tsx`](webview-ui/src/components/chat/ChatView.tsx:1298).

There is also a regression test for the exact boundary shape starting on [`batchNearby.spec.ts`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:270): it uses `[match-1, codebase_search_result, match-2]` on lines [`272`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:272)-[`274`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:274), then asserts the messages remain separate on lines [`282`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:282)-[`285`](webview-ui/src/utils/__tests__/batchNearby.spec.ts:285). So `codebase_search_result` stops batching rather than being bridged.

File: webview-ui/src/utils/batchNearby.ts (hunk 4 above)
Comment (easonLiangWorldedtech) [reply webview-ui/src/utils/batchNearby.ts]: removed

File: webview-ui/src/utils/__tests__/batchNearby.spec.ts [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+
+/** Predicate for realistic qwen tests: matches JSON tool calls. */
+const isToolCall = (m: TestItem) => !!m.text?.startsWith("{")
+
+/** Ignorable: api_req_started/finished, empty text, reasoning. */
+const isIgnorableBetweenTargets = (m: TestItem): boolean => {
+	if (m.type !== "say") return false
+	return (
+		m.say === "api_req_started" ||
+		m.say === "api_req_finished" ||
```
Comment (edelauna) [inline webview-ui/src/utils/__tests__/batchNearby.spec.ts]: The latest commit dropped `api_req_finished` from production `isIgnorableBetweenTargets` (ChatView.tsx:1289), but the fixture here still lists it — and the tests below (~lines 350, 367, 386) rely on it being ignorable. Since it can't reach `batchNearby` in production anyway (`combineApiRequests` merges/drops it and the `visibleMessages` filter strips it again), should we drop it from the fixture too so the tests model the real predicate?

File: webview-ui/src/components/chat/ChatView.tsx [hunk 6]
```diff
@@ -1279,10 +1279,58 @@ const ChatViewComponent: React.ForwardRefRenderFunction<ChatViewRef, ChatViewPro
...
-		const listFilesBatched = batchConsecutive(readFileBatched, isListFilesAsk, synthesizeListFilesBatch)
-		const result = batchConsecutive(listFilesBatched, isEditFileAsk, synthesizeEditFileBatch)
+		// Messages that can be safely skipped over when batching tool asks.
+		// These are low-information or invisible messages that don't affect semantics:
+		// - api_req_started (API request metadata row)
+		// - empty text rows (partial streaming with no visible content)
+		// - reasoning rows (hidden from user by default)
+		const isIgnorableBetweenTargets = (msg: ClineMessage): boolean => {
+			if (msg.type !== "say") return false
+			return msg.say === "api_req_started" || (msg.say === "text" && !msg.text?.trim()) || msg.say === "reasoning"
```
Comment (edelauna) [inline webview-ui/src/components/chat/ChatView.tsx]: These predicates are inline closures, and the spec's copies are hand-synced — which has now drifted twice (the `codebase_search_result` addition, and the `api_req_finished` comment on the spec). Two suggestions:

- Export `isIgnorableBetweenTargets`/`isBoundary` from a small shared module so the spec imports the real predicates instead of copies.
- Add a `ChatView` JSDOM test that feeds `[readFile ask, api_req_started, readFile ask]` and asserts one batched approval row renders instead of two — pins the #1004 regression where it lives (behavior-only per webview-ui/AGENTS.md, so Vitest rather than visual or e2e).

**Review by edelauna — CHANGES_REQUESTED** [review]
Looks good - can you add some additional JSDOM state tests please

**Comment by easonLiangWorldedtech** [comment]
Addressed the review feedback.

Changes made:
- Extracted the batching predicates into a shared utility so production and tests use the same `isIgnorableBetweenTargets` / `isBoundary` logic.
- Updated `ChatView` to import the shared predicates instead of keeping inline closures.
- Updated `batchNearby` tests to import the real predicates and removed the stale `api_req_finished` fixture assumptions, since that row is no longer reachable after `combineApiRequests` / `visibleMessages` filtering.
- Added a JSDOM regression test in `ChatView.spec.tsx` that feeds `[readFile ask, api_req_started, readFile ask]` through the real `ChatView` state path and asserts the readFile asks are batched.

Validation:
- `npx vitest run src/components/chat/__tests__/ChatView.spec.tsx -t "batches readFile asks separated by an API request row"`
- `npx vitest run src/components/chat/__tests__/ChatView.spec.tsx -t "ChatView - Sound Playing Tests"`
- `npx vitest run src/utils/__tests__/batchNearby.spec.ts`
- `pnpm exec eslint --prune-suppressions --max-warnings=0 src/components/chat/ChatView.tsx src/components/chat/__tests__/ChatView.spec.tsx src/utils/__tests__/batchNearby.spec.ts src/utils/chatBatchingPredicates.ts`

The commit also passed the repo pre-commit lint hook.

And also rebased.

**Review by edelauna — APPROVED** [review]
Thanks for addressing the feedback
